internal/ls/utilities.go

```diff
@@ -711,6 +1006,512 @@ func nodeEndsWith(n *ast.Node, expectedLastToken ast.Kind, sourceFile *ast.Sourc
+		}
+		if name == nil || name.Parent.Kind == ast.KindImportEqualsDeclaration {
+			return ast.SemanticMeaningNamespace
+		}
+		return ast.SemanticMeaningAll
+	} else if ast.IsDeclarationName(node) {
+		return getMeaningFromDeclaration(parent)
+	} else if ast.IsEntityName(node) && ast.FindAncestor(node, func(*ast.Node) bool {
+		return node.Kind == ast.KindJSDocNameReference || ast.IsJSDocLinkLike(node) || node.Kind == ast.KindJSDocMemberName
+	}) != nil {
+		return ast.SemanticMeaningAll
+	} else if isTypeReference(node) {
+		return ast.SemanticMeaningType
+	} else if isNamespaceReference(node) {
+		return ast.SemanticMeaningNamespace
+	} else if parent.Kind == ast.KindTypeParameter {
+		// todo jsdoc not implemented
+		// if !ast.IsJSDocTemplateTag(parent.Parent) {
+		// 	panic("should be handled by isDeclarationName")
+		// }
+		return ast.SemanticMeaningType
+	} else if parent.Kind == ast.KindLiteralType {
+		// This might be T["name"], which is actually referencing a property and not a type. So allow both meanings.
+		return ast.SemanticMeaningType | ast.SemanticMeaningValue
+	} else {
+		return ast.SemanticMeaningValue
+	}
+}
+
+func getMeaningFromDeclaration(node *ast.Node) ast.SemanticMeaning {
+	switch node.Kind {
+	case ast.KindVariableDeclaration:
+		// if ast.IsInJSFile(node) { // !!! && ast.GetJSDocEnumTag(node) {
+		// 	return ast.SemanticMeaningAll
+		// }
+		return ast.SemanticMeaningValue
+
+	case ast.KindParameter, ast.KindBindingElement, ast.KindPropertyDeclaration, ast.KindPropertySignature, ast.KindPropertyAssignment, ast.KindShorthandPropertyAssignment, ast.KindMethodDeclaration, ast.KindMethodSignature, ast.KindConstructor, ast.KindGetAccessor, ast.KindSetAccessor, ast.KindFunctionDeclaration, ast.KindFunctionExpression, ast.KindArrowFunction, ast.KindCatchClause, ast.KindJsxAttribute:
+		return ast.SemanticMeaningValue
+
+	case ast.KindTypeParameter, ast.KindInterfaceDeclaration, ast.KindTypeAliasDeclaration, ast.KindTypeLiteral:
+		return ast.SemanticMeaningType
+
+	case ast.KindJSDocTypedefTag:
+		// If it has no name node, it shares the name with the value declaration below it.
+		if node.Name() == nil {
+			return ast.SemanticMeaningValue | ast.SemanticMeaningType
+		}
+		return ast.SemanticMeaningType
+
+	case ast.KindEnumMember, ast.KindClassDeclaration:
+		return ast.SemanticMeaningValue | ast.SemanticMeaningType
+
+	case ast.KindModuleDeclaration:
+		if ast.IsAmbientModule(node) {
+			return ast.SemanticMeaningNamespace | ast.SemanticMeaningValue
+		} else if ast.GetModuleInstanceState(node) == ast.ModuleInstanceStateInstantiated {
+			return ast.SemanticMeaningNamespace | ast.SemanticMeaningValue
+		} else {
+			return ast.SemanticMeaningNamespace
+		}
+
+	case ast.KindEnumDeclaration, ast.KindNamedImports, ast.KindImportSpecifier, ast.KindImportEqualsDeclaration, ast.KindImportDeclaration, ast.KindExportAssignment, ast.KindExportDeclaration:
```

Comment:
```suggestion
	case ast.KindEnumDeclaration, ast.KindNamedImports, ast.KindImportSpecifier, ast.KindImportEqualsDeclaration, ast.KindImportDeclaration, ast.KindJSImportDeclaratoin, ast.KindExportAssignment, ast.KindExportDeclaration:
```

Reply:
does it also need `ast.KindJsExportAssignment`?